File: frontend/src/lib/api/types/workspaceExtensions.ts
/**
 * GrowFlow Workspace Extensions API Types (Batch 06: S27–S33).
 */

export interface GitHubCommitPreview {
  sha: string;
  message: string;
  author: string;
  date: string;
}

export interface GitHubIntegrationResponse {
  id: string | null;
  project_instance_id: string;
  repository_name: string;
  repository_url: string;
  connection_status: 'NOT_CONNECTED' | 'CONNECTED' | 'SYNCING' | 'ERROR';
  default_branch: string;
  commit_count: number;
  last_sync_at: string | null;
  sync_error: string | null;
  cached_commits_preview: GitHubCommitPreview[];
}

export interface GitHubConnectPayload {
  repository_url: string;
  repository_name?: string;
  default_branch?: string;
}

export interface ActivityItemResponse {
  id: string;
  event_type: string;
  title: string;
  description: string;
  actor_role: string;
  actor_id: string | null;
  resource_type: string;
  resource_id: string;
  project_instance_id?: string | null;
  group_id?: string | null;
  occurred_at: string | null;
  metadata: Record<string, any>;
}


export interface AIMentorMessage {
  id: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  sources: Array<{ title: string; section?: string }>;
  suggested_action: {
    action_type: string;
    label: string;
    payload: Record<string, any>;
  } | null;
  created_at: string | null;
}

export interface AIMentorConversationResponse {
  conversation_id: string;
  project_instance_id: string;
  title: string;
  ai_available: boolean;
  messages: AIMentorMessage[];
}

export interface AIMentorSendResponse {
  user_message: AIMentorMessage;
  assistant_message: AIMentorMessage;
  ai_available: boolean;
}

export interface HelpRequestResponse {
  id: string;
  project_instance_id: string;
  student_id: string;
  subject: string;
  description: string;
  category: string;
  priority: 'LOW' | 'MEDIUM' | 'HIGH' | 'URGENT';
  status: 'OPEN' | 'IN_REVIEW' | 'RESOLVED';
  mentor_response: string | null;
  resolved_at: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface HelpRequestCreatePayload {
  subject: string;
  description: string;
  category?: string;
  priority?: string;
}

export interface MentorNoteResponse {
  id: string;
  project_instance_id: string;
  mentor_id: string | null;
  title: string;
  message: string;
  note_type: 'INFORMATIONAL' | 'ACTIONABLE' | 'FEEDBACK';
  status: 'UNREAD' | 'ACKNOWLEDGED';
  related_resource_type: string | null;
  related_resource_id: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface ImpactAnalysis {
  affected_sections: string[];
  affected_tasks_count: number;
  total_tasks_count?: number;
  total_milestones_count?: number;
  estimated_risk: 'LOW' | 'MEDIUM' | 'HIGH';
  analysis_narrative: string;
  duration_impact?: string;
  recommended_action?: string;
}

export interface ProjectChangeResponse {
  id: string;
  project_instance_id: string;
  student_id: string;
  change_title: string;
  change_description: string;
  change_type: 'SCOPE' | 'TECH_STACK' | 'ARCHITECTURE' | 'SCHEDULE';
  status: 'ANALYZED' | 'CONFIRMED' | 'COMPLETED' | 'REJECTED';
  idempotency_key: string | null;
  impact_analysis: ImpactAnalysis;
  source_blueprint_version_number: number;
  resulting_blueprint_version_number: number | null;
  qa_score: number | null;
  qa_feedback: Record<string, any> | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface ProjectChangeAnalyzePayload {
  change_title: string;
  change_description: string;
  change_type: string;
}

export interface ProjectChangeConfirmPayload {
  idempotency_key?: string;
}

export interface BlueprintVersionResponse {
  id: string;
  blueprint_id: string;
  project_instance_id: string;
  version_number: number;
  status: string;
  qa_score: number | null;
  qa_feedback?: Record<string, any> | null;
  change_summary: string | null;
  created_at: string | null;
  content?: Record<string, any> | null;
}

export interface MentorAIChatPayload {
  message: string;
  history?: Array<{ role: string; content: string }>;
}

export interface MentorAIChatResponse {
  user_message: {
    id: string;
    role: string;
    content: string;
    created_at: string;
  };
  assistant_message: {
    id: string;
    role: string;
    content: string;
    sources: Array<{ title: string; section?: string }>;
    created_at: string;
  };
  ai_available: boolean;
  scope: 'GROUP' | 'PORTFOLIO' | string;
  group_id?: string | null;
  group_name?: string | null;
  mentor_id?: string | null;
}

export interface MentorAIStatusResponse {
  ai_available: boolean;
  scope: 'GROUP' | 'PORTFOLIO' | string;
  group_id?: string | null;
  group_name?: string | null;
  mentor_id?: string | null;
}

File: backend/app/api/schemas/workspace_extensions.py
"""
GrowFlow — Workspace Extensions Pydantic Schemas (Batch 06: S27–S33).

Defines request and response models for:
- GitHub observation & sync
- Canonical activity audit trail
- AI Mentor conversations & action confirmation
- Help requests & mentor feedback
- Project changes & versioned regenerations
"""

from __future__ import annotations

from typing import Any
from pydantic import BaseModel, Field


# -----------------------------------------------------------------------------
# GitHub (S27)
# -----------------------------------------------------------------------------
class GitHubConnectPayload(BaseModel):
    repository_url: str = Field(..., min_length=5, description="Full HTTPS URL of GitHub repository")
    repository_name: str | None = Field(None, description="Optional repository name")
    default_branch: str = Field(default="main", description="Primary default branch name")


class GitHubIntegrationResponse(BaseModel):
    id: str | None = None
    project_instance_id: str
    repository_name: str = ""
    repository_url: str = ""
    connection_status: str = "NOT_CONNECTED"
    default_branch: str = "main"
    commit_count: int = 0
    last_sync_at: str | None = None
    sync_error: str | None = None
    cached_commits_preview: list[dict[str, Any]] = []


# -----------------------------------------------------------------------------
# Activity (S28)
# -----------------------------------------------------------------------------
class ActivityItemResponse(BaseModel):
    id: str
    event_type: str
    title: str
    description: str
    actor_role: str
    actor_id: str | None = None
    resource_type: str
    resource_id: str
    occurred_at: str | None = None
    metadata: dict[str, Any] = {}


# -----------------------------------------------------------------------------
# AI Mentor (S29)
# -----------------------------------------------------------------------------
class AIMentorSendMessagePayload(BaseModel):
    content: str = Field(..., min_length=1, max_length=4000)


class AIMentorExecuteActionPayload(BaseModel):
    action_type: str = Field(..., description="Action type to execute (e.g. CREATE_TASK, CREATE_HELP_REQUEST)")
    payload: dict[str, Any] = Field(default_factory=dict)


class AIMentorMessageResponse(BaseModel):
    id: str
    role: str
    content: str
    sources: list[dict[str, Any]] = []
    suggested_action: dict[str, Any] | None = None
    created_at: str | None = None


class AIMentorConversationResponse(BaseModel):
    conversation_id: str
    project_instance_id: str
    title: str
    ai_available: bool
    messages: list[AIMentorMessageResponse] = []


class AIMentorSendResponse(BaseModel):
    user_message: AIMentorMessageResponse
    assistant_message: AIMentorMessageResponse
    ai_available: bool


# -----------------------------------------------------------------------------
# Help Requests (S30)
# -----------------------------------------------------------------------------
class HelpRequestCreatePayload(BaseModel):
    subject: str = Field(..., min_length=3, max_length=255)
    description: str = Field(..., min_length=5, max_length=5000)
    category: str = Field(default="TECHNICAL")
    priority: str = Field(default="MEDIUM")


class HelpRequestResponse(BaseModel):
    id: str
    project_instance_id: str
    student_id: str
    subject: str
    description: str
    category: str
    priority: str
    status: str
    mentor_response: str | None = None
    resolved_at: str | None = None
    created_at: str | None = None
    updated_at: str | None = None


# -----------------------------------------------------------------------------
# Mentor Feedback (S31)
# -----------------------------------------------------------------------------
class MentorNoteResponse(BaseModel):
    id: str
    project_instance_id: str
    mentor_id: str | None = None
    title: str
    message: str
    note_type: str
    status: str
    related_resource_type: str | None = None
    related_resource_id: str | None = None
    created_at: str | None = None
    updated_at: str | None = None


# -----------------------------------------------------------------------------
# Project Changes & Regeneration (S32 & S33)
# -----------------------------------------------------------------------------
class ProjectChangeAnalyzePayload(BaseModel):
    change_title: str = Field(..., min_length=3, max_length=255)
    change_description: str = Field(..., min_length=10, max_length=5000)
    change_type: str = Field(default="SCOPE")  # SCOPE, TECH_STACK, ARCHITECTURE, SCHEDULE


class ProjectChangeConfirmPayload(BaseModel):
    idempotency_key: str | None = Field(None, max_length=100)


class ProjectChangeResponse(BaseModel):
    id: str
    project_instance_id: str
    student_id: str
    change_title: str
    change_description: str
    change_type: str
    status: str
    idempotency_key: str | None = None
    impact_analysis: dict[str, Any] = {}
    source_blueprint_version_number: int = 1
    resulting_blueprint_version_number: int | None = None
    qa_score: int | None = None
    qa_feedback: dict[str, Any] | None = None
    created_at: str | None = None
    updated_at: str | None = None


class BlueprintVersionResponse(BaseModel):
    id: str
    blueprint_id: str
    project_instance_id: str
    version_number: int
    status: str
    qa_score: int | None = None
    qa_feedback: dict[str, Any] | None = None
    change_summary: str | None = None
    created_at: str | None = None
    content: dict[str, Any] | None = None
